Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,297 @@
+// becomes available. It correctly handles entries being asynchonously added or
+// removed from the schedule.
+//
+// The details of time management are completely concealed within the class.
+// Once an entry is scheduled, there is no way to reschedule or even retrieve
+// the time.
+template <typename T, typename DurationT = std::chrono::system_clock::duration>
+class Schedule {
+  // Internal invariants:
+  // - entries are always in sorted order, leftmost entry is always the most
+  //   due;
+  // - each operation modifying the queue notifies the condition variable `cv_`.
+ public:
+  using Duration = DurationT;
+  // Entries are scheduled using absolute time.
+  using TimePoint =
+      std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+  // Schedules an entry for the specified time due. `due` may be in the past.
+  void Push(const T& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{value, due});
+  }
+  void Push(T&& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{std::move(value), due});
+  }
+
+  // If the queue contains at least one entry for which the scheduled time is
+  // due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue, moves it into `out`, and returns true. If no
+  // entry is due, doesn't modify `out` and returns false. `out` may be
+  // `nullptr` (in which case the value will be simply discarded).
+  bool PopIfDue(T* const out) {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    if (HasDue()) {
+      DoPop(out, scheduled_.begin());
+      return true;
+    }
+    return false;
+  }
+
+  // Removes the first entry satisfying predicate from the queue, moves it into
+  // `out`, and returns true. If no such entry exists, doesn't modify `out` and
+  // returns false. Predicate is applied to entries in order according to their
+  // scheduled time. `out` may be `nullptr` (in which case the value will be
+  // simply discarded).
+  //
+  // Note that this function doesn't take into account whether the removed entry
+  // is past its due time.
+  template <typename Pred>
+  bool RemoveIf(T* const out, const Pred pred) {
+    std::lock_guard<std::mutex> lock{mutex_};
+    const auto found =
+        std::find_if(scheduled_.begin(), scheduled_.end(),
+                     [&pred](const Entry& s) { return pred(s.value); });
+    if (found != scheduled_.end()) {
+      DoPop(out, found);
+      return true;
+    }
+    return false;
+  }
+
+  // Blocks until at least one entry is available for which the scheduled time
+  // is due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue and moves it into `out`. The function will
+  // attempt to minimize both the waiting time and busy waiting. `out` may be
+  // `nullptr` (in which case the value will be simply discarded).
+  void PopBlocking(T* const out) {
+    std::unique_lock<std::mutex> lock{mutex_};
```

Comment:
There's a lot of work done here. Consider moving it out to an implementation file.

http://go/cpp-style#Inline_Functions